src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenForEachTests.cs

```diff
@@ -2365,5 +2365,696 @@ .locals init (T V_0)
+                );
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaExtensionWithOptionalParameter()
+        {
+            string source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.WriteLine(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public Enumerator(int start) => Current = start;
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this C self, int x = 1) => new C.Enumerator(x);
+}";
+            var comp = CreateCompilationWithMscorlib46(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: @"2
+3");
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaExtensionWithParams()
+        {
+            string source = @"
+using System;
+public class C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.WriteLine(i);
+        }
+    }
+    public sealed class Enumerator
+    {
+        public Enumerator(int[] arr) => Current = arr.Length;
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this C self, params int[] x) => new C.Enumerator(x);
+}";
+            var comp = CreateCompilationWithMscorlib46(source, options: TestOptions.DebugExe, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: @"1
+2
+3");
+        }
+
+        [Fact]
+        public void TestGetEnumeratorPatternViaRefExtensionOnNonAssignableVariable()
+        {
+            // See https://github.com/dotnet/csharplang/blob/98043cdc889303d956d540d7ab3bc4f5044a9d3b/meetings/2018/LDM-2018-01-24.md#deconstruct-as-ref-extension-method
+            // I imagine same decision should apply here
+            string source = @"
+using System;
+public struct C
+{
+    public static void Main()
+    {
+        foreach (var i in new C())
+        {
+            Console.WriteLine(i);
+        }
+    }
+    public struct Enumerator
+    {
+        public int Current { get; private set; }
+        public bool MoveNext() => Current++ != 3;
+    }
+}
+public static class Extensions
+{
+    public static C.Enumerator GetEnumerator(this ref C self) => new C.Enumerator();
+}";
+            var comp = CreateCompilationWithMscorlib46(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (7,27): error CS1510: A ref or out value must be an assignable variable
```

Comment:
This is not a great error, but:
a) It's the same one as for Deconstruct
b) I doubt anyone's likely to hit this in practice so not worth creating a custom error message till someone complains